### PR TITLE
[KYUUBI #1577] Add DropIgnoreNonexistent Rule.

### PR DESCRIPTION
**Pull request by wForget**
### _Why are the changes needed?_

Like the `hive.exec.drop.ignorenonexistent` configuration in hive, do not report an error if DROP DATABASE/TABLE/VIEW/Function/Partition specifies a non-existent database/table/view/function/partition. For details: #1577.

### _How was this patch tested?_
- [X] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [X] [Run test](https://kyuubi.readthedocs.io/en/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/DropIgnoreNonexistent.scala [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+import org.apache.kyuubi.sql.KyuubiSQLConf._
+
+case class DropIgnoreNonexistent(session: SparkSession) extends Rule[LogicalPlan] {
+
+  override def apply(plan: LogicalPlan): LogicalPlan = {
+    if (conf.getConf(DROP_IGNORE_NONEXISTENT)) {
+      plan match {
+        case i @ AlterTableDropPartitionCommand(tableName, specs, ifExists, purge, retainData)
+          if !ifExists =>
```
Comment (ulysses-you) [inline dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/DropIgnoreNonexistent.scala:32]: `two more indention`

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/DropIgnoreNonexistent.scala [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+          i.copy(tableName, specs, true, purge, retainData)
+        case i @ DropTable(child, ifExists, purge) if !ifExists =>
```
Comment (ulysses-you) [inline dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/DropIgnoreNonexistent.scala:34]: Why we need catch `DropTable` ?

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/DropIgnoreNonexistent.scala [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+import org.apache.kyuubi.sql.KyuubiSQLConf._
+
+case class DropIgnoreNonexistent(session: SparkSession) extends Rule[LogicalPlan] {
+
+  override def apply(plan: LogicalPlan): LogicalPlan = {
+    if (conf.getConf(DROP_IGNORE_NONEXISTENT)) {
+      plan match {
+        case i @ AlterTableDropPartitionCommand(tableName, specs, ifExists, purge, retainData)
+          if !ifExists =>
+          i.copy(tableName, specs, true, purge, retainData)
```
Comment (ulysses-you) [inline dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/DropIgnoreNonexistent.scala:33]: `i.copy(ifExists = true)`

File: dev/kyuubi-extension-spark-common/src/main/scala/org/apache/kyuubi/sql/KyuubiSQLConf.scala [hunk 4]
```diff
@@ -114,4 +114,12 @@ object KyuubiSQLConf {
       .version("1.4.0")
       .intConf
       .createOptional
+
+  val DROP_IGNORE_NONEXISTENT =
+    buildConf("spark.sql.analyzer.drop.ignorenonexistent")
```
Comment (ulysses-you) [inline dev/kyuubi-extension-spark-common/src/main/scala/org/apache/kyuubi/sql/KyuubiSQLConf.scala:119]: `spark.sql.optimizer.dropIgnoreNonExistent `

**Review by ulysses-you — COMMENTED** [review]
thank you @wForget , please update the new configuration in `docs/sql/rules.md`.

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/DropIgnoreNonexistent.scala (hunk 1 above)
Comment (wForget) [reply dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/DropIgnoreNonexistent.scala:32]: > `two more indention`

There seems to be no problem with the indentation here.

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/DropIgnoreNonexistent.scala (hunk 2 above)
Comment (wForget) [reply dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/DropIgnoreNonexistent.scala:34]: > Why we need catch `DropTable` ?

When the table does not exist, it does not seem to be resolved as a DropTableCommand.

**Comment by wForget** [comment]
> thank you @wForget , please update the new configuration in `docs/sql/rules.md`.

Hi @ulysses-you , do I need to modify it manually? I didn't find the test case that generated it.

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/DropIgnoreNonexistent.scala (hunk 1 above)
Comment (wForget) [reply dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/DropIgnoreNonexistent.scala:32]: > There seems to be no problem with the indentation here.

Sorry, please ignore it

**Comment by ulysses-you** [comment]
> do I need to modify it manually? I didn't find the test case that generated it.

@wForget yes, it's a manual work : (

File: dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/DropIgnoreNonexistent.scala (hunk 2 above)
Comment (ulysses-you) [reply dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/DropIgnoreNonexistent.scala]: ah I see. It seems not compatiable between Spark 3.1 and 3.2.
- For 3.1, `DropTable` ifExists -> `NoopDropTable `
- For 3.2, `DropTable` ifExists -> `ResolveCommandsWithIfExists`

So  it looks more better to me if we can add this rule using `injectPostHocResolutionRule` like `NoopDropTable` or `ResolveCommandsWithIfExists`. What do you think ?

Comment (wForget) [reply dev/kyuubi-extension-spark-3-1/src/main/scala/org/apache/kyuubi/sql/DropIgnoreNonexistent.scala]: Thanks @ulysses-you , it sounds good to me, I will try it.

**Review by ulysses-you — APPROVED** [review]

**Comment by ulysses-you** [comment]
thanks, merging to master

**Pull request closed without merging**